File: Algorithm/DP/Fibonacci_series.cpp
/*
problem statement :
Given a number n, display the Fibonacci sequence of first n numbers

The Fibonacci sequence is a series where the next term is the sum of pervious two terms. The first two terms of the Fibonacci sequence is 0 followed by 1.
The Fibonacci sequence: 0, 1, 1, 2, 3, 5, 8, 13, 21....

Approach : The sequence F(n) of Fibonacci numbers is defined by the recurrence relation

F(n) = F(n - 1) + F(n - 2)
with seed values ,

F(0) = 0 and F(1) = 1.

for the n = 4 , the recursive tree would be ,

               fib(4)
            /        \
         fib(3)      fib(2)
        /    \       /    \
  fib(2)   fib(1)  fib(1) fib(0)
  /     \
fib(1) fib(0)

*/

// code part (Method 1 : using recursion)


#include<bits/stdc++.h>
using namespace std;

//Fibonacci Series using Recursion

int fib(int n)
{
	// base case : if n == 0 ,then return 0 and if n == 1 , then return 1
	if(n <= 1)
	{
		return n;
	}

	// recursive call
	return fib(n - 1) + fib(n - 2);

}
int main()
{
	int n;

	cout << "Enter the number of elements:";

	cin >> n;

// printing the fibonacci sequence

	for(int i = 0 ; i < n; i++)
	{
		cout << fib(i) << " ";

	}

	cout << endl;

	return 0;
}

// code part (Method 2 : using dynamic programming)

#include<bits/stdc++.h>
using namespace std;

// Fibonacci series using dynamic progarmming

void fib(int n) {

// let's make an array to store our computed results
	int dp[n];


	dp[0] = 0;

	dp[1] = 1;

// calculating the remaining values

	for (int i = 2; i < n; i++)
	{
		dp[i] = dp[i - 1] + dp[i - 2];
	}

// displaying the fibonacci series

	for (int i = 0; i < n; i++)
	{
		cout << dp[i] << " ";
	}
}

int main()
{
	int n;

	cout << "Enter the number of elements:";

	cin >> n;

    fib(n);

	return 0;
}


/*
(For method 1)

// Time Complexity : T(n) = T(n-1) + T(n-2) which is exponential.
// Space complexity : O(n) , if we consider stack space in recursive calls


*/

/*

(For method 2)

// Time Complexity : O(n)
// Space Complexity : O(1)

*/
/*
Sample test cases


Enter the number of elements: 7
0 1 1 2 3 5 8

Enter the number of elements: 10
0 1 1 2 3 5 8 13 21 34


*/

File: Code/Java/Fibonacci_Series.java
/*The Fibonacci sequence is a series where the next term is the sum of pervious two terms. The first two terms of the Fibonacci sequence is 0 followed by 1.
The Fibonacci sequence: 0, 1, 1, 2, 3, 5, 8, 13, 21....
*/
package fibonaccinumbers;
import java.util.Scanner;

/*Method 1 : Using For Loop*/
class FibonacciExample1{  
          public static void main(String args[])  
             {   
                  Scanner sc = new Scanner(System.in);
                  System.out.println("Enter count : ");
                  int count = sc.nextInt();
                  System.out.println(count);
                  int t1 = 0;
                  int t2 = 1;
                  int t3;
                  int i;    
                  System.out.print(t1 +" "+ t2);        //printing 0 and 1    
    
                  for(i = 2 ; i < count ; i++) {               //loop starts from 2 because 0 and 1 are already printed        
                        t3 = t1 + t2;    
                        System.out.print(" "+t3);    
                        t1 = t2;    
                        t2 = t3;    
                    }    
  
          }
}


/*Method 2: Using Recursion*/
class FibonacciExample2{  
    public static int t1 = 0;
    public static int t2 = 1;
    public static int t3 = 0;    
    public static void printFibonacci(int count){    
       if(count>0){    
            t3 = t1 + t2;    
            t1 = t2;    
            t2 = t3;    
            System.out.print(" " + t3);   
            printFibonacci(count-1);    
        }    
    }    
    public static void main(String args[]){    
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter count : ");
        int count = sc.nextInt();   
        System.out.print(t1+" "+t2);                    //printing 0 and 1    
        printFibonacci(count-2);                        //n-2 because 2 numbers are already printed   
    }  
 }  

/*Method 3: Using Dynamic Programming*/
class FibonacciExample3
{
  public static int fibonacci(int n)
    {
    /* Declare an array to store Fibonacci numbers. */
          int arr[] = new int[n+2];   // 1 extra to handle case, n = 0
          int i;
      
    /* 0th and 1st number of the series are 0 and 1*/
          arr[0] = 0;
          arr[1] = 1;
     
         for (i = 2; i <= n; i++) {
       /* Add the previous 2 numbers in the series and store it */
              arr[i] = arr[i-1] + arr[i-2];
           }
      
     return arr[n];
 }
      
    public static void main (String args[])
    {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter count : ");
        int count = sc.nextInt();
        System.out.println(fibonacci(count));
    }
}
/*
 
  Test case 1
  Input : Enter the number of elements: 6
  Output : 0 1 1 2 3 5
  Test case 2
  Input : Enter the number of elements: 9
  Output : 0 1 1 2 3 5 8 13 21
   
  Time Complexity and Space Complexity
  (Method 1 - Via for Loop)
  Time Complexity : O(n)
  Space Complexity :O(1)
   
  (Method 2 - Via Recursion)
  Time Complexity : T(n) = T(n-1) + T(n-2) which is exponential.
  Space complexity : O(n) , if we consider stack space in recursive calls
 
  (Method 3 : Via Dyanamic Programing)
  Time Complexity : O(n)
  Space Complexity : O(1)
 */ 
